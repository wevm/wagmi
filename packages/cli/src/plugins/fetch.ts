import { mkdir, readFile, writeFile } from 'node:fs/promises'
import { homedir } from 'node:os'
import { join } from 'pathe'

import type { Abi } from 'viem'
import type { ContractConfig, Plugin } from '../config.js'
import type { Compute, RequiredBy } from '../types.js'

export type FetchConfig = {
  /**
   * Duration in milliseconds to cache ABIs from request.
   *
   * @default 1_800_000 // 30m in ms
   */
  cacheDuration?: number | undefined
  /**
   * Contracts to fetch ABIs for.
   */
  contracts: Compute<Omit<ContractConfig, 'abi'>>[]
  /**
   * Function for creating a cache key for contract.
   */
  getCacheKey?:
    | ((config: { contract: Compute<Omit<ContractConfig, 'abi'>> }) => string)
    | undefined
  /**
   * Name of source.
   */
  name?: ContractConfig['name'] | undefined
  /**
   * Function for parsing ABI from fetch response.
   *
   * @default ({ response }) => response.json()
   */
  parse?:
    | ((config: {
        response: Response
      }) => ContractConfig['abi'] | Promise<ContractConfig['abi']>)
    | undefined
  /**
   * Function for returning a request to fetch ABI from.
   */
  request: (config: {
    address?: ContractConfig['address'] | undefined
    name: ContractConfig['name']
  }) =>
    | { url: RequestInfo; init?: RequestInit | undefined }
    | Promise<{ url: RequestInfo; init?: RequestInit | undefined }>
  /**
   * Duration in milliseconds before request times out.
   *
   * @default 5_000 // 5s in ms
   */
  timeoutDuration?: number | undefined
}

type FetchResult = Compute<RequiredBy<Plugin, 'contracts'>>

/** Fetches and parses contract ABIs from network resource with `fetch`. */
export function fetch(config: FetchConfig): FetchResult {
  const {
    cacheDuration = 1_800_000,
    contracts: contractConfigs,
    getCacheKey = ({ contract }) => JSON.stringify(contract),
    name = 'Fetch',
    parse = ({ response }) => response.json(),
    request,
    timeoutDuration = 5_000,
  } = config

  return {
    async contracts() {
      const cacheDir = getCacheDir()
      await mkdir(cacheDir, { recursive: true })

      const timestamp = Date.now() + cacheDuration
      const contracts = []
      for (const contract of contractConfigs) {
        const cacheKey = getCacheKey({ contract })
        const cacheFilePath = join(cacheDir, `${cacheKey}.json`)
        const cachedFile = JSON.parse(
          await readFile(cacheFilePath, 'utf8').catch(() => 'null'),
        )

        let abi: Abi | undefined
        if (cachedFile?.timestamp > Date.now()) abi = cachedFile.abi
        else {
          try {
            const controller = new globalThis.AbortController()
            const timeout = setTimeout(
              () => controller.abort(),
              timeoutDuration,
            )

            const { url, init } = await request(contract)
            const response = await globalThis.fetch(url, {
              ...init,
              signal: controller.signal,
            })
            clearTimeout(timeout)

            abi = await parse({ response })
            await writeFile(
              cacheFilePath,
              `${JSON.stringify({ abi, timestamp }, undefined, 2)}\n`,
            )
          } catch (error) {
            try {
              // Attempt to read from cache if fetch fails.
              abi = JSON.parse(await readFile(cacheFilePath, 'utf8')).abi
            } catch {}
            if (!abi) throw error
          }
        }

        if (!abi) throw Error('Failed to fetch ABI for contract.')
        contracts.push({ abi, address: contract.address, name: contract.name })
      }
      return contracts
    },
    name,
  }
}

export function getCacheDir() {
  return join(homedir(), '.wagmi-cli/plugins/fetch/cache')
}
