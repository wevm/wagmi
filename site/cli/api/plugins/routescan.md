# routescan

Plugin for fetching ABIs from [Routescan](https://routescan.io) and adding into `contracts` config.

## Import

```ts
import { routescan } from '@wagmi/cli/plugins'
```

## Usage

```ts{2,6-14}
import { defineConfig } from '@wagmi/cli'
import { routescan } from '@wagmi/cli/plugins'

export default defineConfig({
  plugins: [
    routescan({
      apiKey: process.env.ROUTESCAN_API_KEY,
      chainId: 1,
      contracts: [
        {
          name: 'Wagmigotchi',
          address: '0xecb504d39723b0be0e3a9aa33d646642d1051ee1',
        },
      ],
    }),
  ],
})
```

## Configuration

```ts
import { type RoutescanConfig } from '@wagmi/cli/plugins'
```

### apiKey

`string`

Routescan API key. Routescan API keys are specific per network and include testnets (e.g. Ethereum Mainnet and Sepolia share same API key).

```ts
import { defineConfig } from '@wagmi/cli'
import { routescan } from '@wagmi/cli/plugins'

export default defineConfig({
  plugins: [
    routescan({
      apiKey: process.env.ROUTESCAN_API_KEY, // [!code focus]
      chainId: 1,
      contracts: [
        {
          name: 'Wagmigotchi',
          address: '0xecb504d39723b0be0e3a9aa33d646642d1051ee1',
        },
      ],
    }),
  ],
})
```

### cacheDuration

`number | undefined`

- Duration in milliseconds to cache ABIs.
- Defaults to `1_800_000` (30 minutes).

```ts
import { defineConfig } from '@wagmi/cli'
import { routescan } from '@wagmi/cli/plugins'

export default defineConfig({
  plugins: [
    routescan({
      apiKey: process.env.ROUTESCAN_API_KEY,
      cacheDuration: 300_000, // [!code focus]
      chainId: 1,
      contracts: [
        {
          name: 'Wagmigotchi',
          address: '0xecb504d39723b0be0e3a9aa33d646642d1051ee1',
        },
      ],
    }),
  ],
})
```

### chainId

`number`

Chain ID to use for fetching ABI. If [`address`](/cli/config/options#address) is an object, `chainId` is used to select the address.

```ts
import { defineConfig } from '@wagmi/cli'
import { blockExplorer } from '@wagmi/cli/plugins'

export default defineConfig({
  plugins: [
    routescan({
      apiKey: process.env.ROUTESCAN_API_KEY,
      chainId: 1, // [!code focus]
      contracts: [
        {
          name: 'Wagmigotchi',
          address: '0xecb504d39723b0be0e3a9aa33d646642d1051ee1',
        },
        {
          name: 'EnsRegistry',
          address: {
            1: '0x314159265dd8dbb310642f98f50c066173c1259b',
            5: '0x112234455c3a32fd11230c42e7bccd4a84e02010',
          },
        },
      ],
    }),
  ],
})
```

### contracts

`{ name: string; address?: Address | Record<number, Address> | undefined }[]`

Contracts to fetch ABIs for.

```ts
import { defineConfig } from '@wagmi/cli'
import { routescan } from '@wagmi/cli/plugins'

export default defineConfig({
  plugins: [
    routescan({
      apiKey: process.env.ROUTESCAN_API_KEY,
      chainId: 1,
      contracts: [ // [!code focus]
        { // [!code focus]
          name: 'Wagmigotchi', // [!code focus]
          address: '0xecb504d39723b0be0e3a9aa33d646642d1051ee1', // [!code focus]
        }, // [!code focus]
      ], // [!code focus]
    }),
  ],
})
```

### tryFetchProxyImplementation

`boolean | undefined`

- Whether to try fetching proxy implementation address of the contract.
- Defaults to `false`.

```ts
import { defineConfig } from '@wagmi/cli'
import { blockExplorer } from '@wagmi/cli/plugins'

export default defineConfig({
  plugins: [
    routescan({
      apiKey: process.env.ROUTESCAN_API_KEY,
      chainId: 1, 
      contracts: [
        {
          name: 'FiatToken',
          address: '0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48',
        },
      ],
      tryFetchProxyImplementation: true, // [!code focus]
    }),
  ],
})
```


