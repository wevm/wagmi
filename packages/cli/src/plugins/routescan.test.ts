import { mkdir, rm } from 'node:fs/promises'
import { setupServer } from 'msw/node'
import { afterAll, afterEach, beforeAll, expect, test } from 'vitest'

import {
  address,
  apiKey,
  getHandlers,
  invalidApiKey,
  proxyAddress,
  timeoutAddress,
  unverifiedContractAddress,
} from '../../test/utils.js'
import { getCacheDir } from './fetch.js'
import { routescan } from './routescan.js'

const server = setupServer(...getHandlers('https://api.routescan.io/v2/api'))

beforeAll(() => server.listen())
afterEach(() => server.resetHandlers())
afterAll(() => server.close())

test('fetches ABI', async () => {
  await expect(
    routescan({
      apiKey,
      chainId: 1,
      contracts: [{ name: 'WagmiMintExample', address }],
    }).contracts?.(),
  ).resolves.toMatchSnapshot()
})

test('fetches ABI with multichain deployment', async () => {
  await expect(
    routescan({
      apiKey,
      chainId: 1,
      contracts: [
        { name: 'WagmiMintExample', address: { 1: address, 10: address } },
      ],
    }).contracts?.(),
  ).resolves.toMatchSnapshot()
})

test('fails to fetch for unverified contract', async () => {
  await expect(
    routescan({
      apiKey,
      chainId: 1,
      contracts: [
        { name: 'WagmiMintExample', address: unverifiedContractAddress },
      ],
    }).contracts?.(),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    '[Error: Contract source code not verified]',
  )
})

test('missing address for chainId', async () => {
  await expect(
    routescan({
      apiKey,
      chainId: 1,
      // @ts-expect-error `chainId` and `keyof typeof contracts[number].address` mismatch
      contracts: [{ name: 'WagmiMintExample', address: { 10: address } }],
    }).contracts?.(),
  ).rejects.toThrowErrorMatchingInlineSnapshot(
    `[Error: No address found for chainId "1". Make sure chainId "1" is set as an address.]`,
  )
})

test('invalid api key', async () => {
  await expect(
    routescan({
      apiKey: invalidApiKey,
      chainId: 1,
      contracts: [{ name: 'WagmiMintExample', address: timeoutAddress }],
    }).contracts?.(),
  ).rejects.toThrowErrorMatchingInlineSnapshot('[Error: Invalid API Key]')
})

test('tryFetchProxyImplementation: fetches ABI', async () => {
  const cacheDir = getCacheDir()
  await mkdir(cacheDir, { recursive: true })

  await expect(
    routescan({
      apiKey,
      chainId: 1,
      contracts: [{ name: 'WagmiMintExample', address }],
      tryFetchProxyImplementation: true,
    }).contracts?.(),
  ).resolves.toMatchSnapshot()

  await rm(cacheDir, { recursive: true })
})

test('tryFetchProxyImplementation: fetches implementation ABI', async () => {
  const cacheDir = getCacheDir()
  await mkdir(cacheDir, { recursive: true })

  await expect(
    routescan({
      apiKey,
      chainId: 1,
      contracts: [{ name: 'FiatToken', address: proxyAddress }],
      tryFetchProxyImplementation: true,
    }).contracts?.(),
  ).resolves.toMatchSnapshot()

  await rm(cacheDir, { recursive: true })
})
