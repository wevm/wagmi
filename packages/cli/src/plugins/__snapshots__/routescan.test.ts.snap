// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`fetches ABI 1`] = `
[
  {
    "abi": [
      {
        "inputs": [],
        "stateMutability": "nonpayable",
        "type": "constructor",
      },
      {
        "anonymous": false,
        "inputs": [
          {
            "indexed": true,
            "internalType": "address",
            "name": "owner",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "address",
            "name": "approved",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "Approval",
        "type": "event",
      },
      {
        "anonymous": false,
        "inputs": [
          {
            "indexed": true,
            "internalType": "address",
            "name": "owner",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "address",
            "name": "operator",
            "type": "address",
          },
          {
            "indexed": false,
            "internalType": "bool",
            "name": "approved",
            "type": "bool",
          },
        ],
        "name": "ApprovalForAll",
        "type": "event",
      },
      {
        "anonymous": false,
        "inputs": [
          {
            "indexed": true,
            "internalType": "address",
            "name": "from",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "Transfer",
        "type": "event",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "approve",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "owner",
            "type": "address",
          },
        ],
        "name": "balanceOf",
        "outputs": [
          {
            "internalType": "uint256",
            "name": "",
            "type": "uint256",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "getApproved",
        "outputs": [
          {
            "internalType": "address",
            "name": "",
            "type": "address",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "owner",
            "type": "address",
          },
          {
            "internalType": "address",
            "name": "operator",
            "type": "address",
          },
        ],
        "name": "isApprovedForAll",
        "outputs": [
          {
            "internalType": "bool",
            "name": "",
            "type": "bool",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "mint",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "name",
        "outputs": [
          {
            "internalType": "string",
            "name": "",
            "type": "string",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "ownerOf",
        "outputs": [
          {
            "internalType": "address",
            "name": "",
            "type": "address",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "from",
            "type": "address",
          },
          {
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "safeTransferFrom",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "from",
            "type": "address",
          },
          {
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
          {
            "internalType": "bytes",
            "name": "_data",
            "type": "bytes",
          },
        ],
        "name": "safeTransferFrom",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "operator",
            "type": "address",
          },
          {
            "internalType": "bool",
            "name": "approved",
            "type": "bool",
          },
        ],
        "name": "setApprovalForAll",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "bytes4",
            "name": "interfaceId",
            "type": "bytes4",
          },
        ],
        "name": "supportsInterface",
        "outputs": [
          {
            "internalType": "bool",
            "name": "",
            "type": "bool",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "symbol",
        "outputs": [
          {
            "internalType": "string",
            "name": "",
            "type": "string",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "tokenURI",
        "outputs": [
          {
            "internalType": "string",
            "name": "",
            "type": "string",
          },
        ],
        "stateMutability": "pure",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "totalSupply",
        "outputs": [
          {
            "internalType": "uint256",
            "name": "",
            "type": "uint256",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "from",
            "type": "address",
          },
          {
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "transferFrom",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
    ],
    "address": {
      "1": "0xaf0326d92b97df1221759476b072abfd8084f9be",
    },
    "name": "WagmiMintExample",
  },
]
`;

exports[`fetches ABI with multichain deployment 1`] = `
[
  {
    "abi": [
      {
        "inputs": [],
        "stateMutability": "nonpayable",
        "type": "constructor",
      },
      {
        "anonymous": false,
        "inputs": [
          {
            "indexed": true,
            "internalType": "address",
            "name": "owner",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "address",
            "name": "approved",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "Approval",
        "type": "event",
      },
      {
        "anonymous": false,
        "inputs": [
          {
            "indexed": true,
            "internalType": "address",
            "name": "owner",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "address",
            "name": "operator",
            "type": "address",
          },
          {
            "indexed": false,
            "internalType": "bool",
            "name": "approved",
            "type": "bool",
          },
        ],
        "name": "ApprovalForAll",
        "type": "event",
      },
      {
        "anonymous": false,
        "inputs": [
          {
            "indexed": true,
            "internalType": "address",
            "name": "from",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "Transfer",
        "type": "event",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "approve",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "owner",
            "type": "address",
          },
        ],
        "name": "balanceOf",
        "outputs": [
          {
            "internalType": "uint256",
            "name": "",
            "type": "uint256",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "getApproved",
        "outputs": [
          {
            "internalType": "address",
            "name": "",
            "type": "address",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "owner",
            "type": "address",
          },
          {
            "internalType": "address",
            "name": "operator",
            "type": "address",
          },
        ],
        "name": "isApprovedForAll",
        "outputs": [
          {
            "internalType": "bool",
            "name": "",
            "type": "bool",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "mint",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "name",
        "outputs": [
          {
            "internalType": "string",
            "name": "",
            "type": "string",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "ownerOf",
        "outputs": [
          {
            "internalType": "address",
            "name": "",
            "type": "address",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "from",
            "type": "address",
          },
          {
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "safeTransferFrom",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "from",
            "type": "address",
          },
          {
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
          {
            "internalType": "bytes",
            "name": "_data",
            "type": "bytes",
          },
        ],
        "name": "safeTransferFrom",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "operator",
            "type": "address",
          },
          {
            "internalType": "bool",
            "name": "approved",
            "type": "bool",
          },
        ],
        "name": "setApprovalForAll",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "bytes4",
            "name": "interfaceId",
            "type": "bytes4",
          },
        ],
        "name": "supportsInterface",
        "outputs": [
          {
            "internalType": "bool",
            "name": "",
            "type": "bool",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "symbol",
        "outputs": [
          {
            "internalType": "string",
            "name": "",
            "type": "string",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "tokenURI",
        "outputs": [
          {
            "internalType": "string",
            "name": "",
            "type": "string",
          },
        ],
        "stateMutability": "pure",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "totalSupply",
        "outputs": [
          {
            "internalType": "uint256",
            "name": "",
            "type": "uint256",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "from",
            "type": "address",
          },
          {
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "transferFrom",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
    ],
    "address": {
      "1": "0xaf0326d92b97df1221759476b072abfd8084f9be",
      "10": "0xaf0326d92b97df1221759476b072abfd8084f9be",
    },
    "name": "WagmiMintExample",
  },
]
`;

exports[`tryFetchProxyImplementation: fetches ABI 1`] = `
[
  {
    "abi": [
      {
        "inputs": [],
        "stateMutability": "nonpayable",
        "type": "constructor",
      },
      {
        "anonymous": false,
        "inputs": [
          {
            "indexed": true,
            "internalType": "address",
            "name": "owner",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "address",
            "name": "approved",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "Approval",
        "type": "event",
      },
      {
        "anonymous": false,
        "inputs": [
          {
            "indexed": true,
            "internalType": "address",
            "name": "owner",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "address",
            "name": "operator",
            "type": "address",
          },
          {
            "indexed": false,
            "internalType": "bool",
            "name": "approved",
            "type": "bool",
          },
        ],
        "name": "ApprovalForAll",
        "type": "event",
      },
      {
        "anonymous": false,
        "inputs": [
          {
            "indexed": true,
            "internalType": "address",
            "name": "from",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "indexed": true,
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "Transfer",
        "type": "event",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "approve",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "owner",
            "type": "address",
          },
        ],
        "name": "balanceOf",
        "outputs": [
          {
            "internalType": "uint256",
            "name": "",
            "type": "uint256",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "getApproved",
        "outputs": [
          {
            "internalType": "address",
            "name": "",
            "type": "address",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "owner",
            "type": "address",
          },
          {
            "internalType": "address",
            "name": "operator",
            "type": "address",
          },
        ],
        "name": "isApprovedForAll",
        "outputs": [
          {
            "internalType": "bool",
            "name": "",
            "type": "bool",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "mint",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "name",
        "outputs": [
          {
            "internalType": "string",
            "name": "",
            "type": "string",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "ownerOf",
        "outputs": [
          {
            "internalType": "address",
            "name": "",
            "type": "address",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "from",
            "type": "address",
          },
          {
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "safeTransferFrom",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "from",
            "type": "address",
          },
          {
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
          {
            "internalType": "bytes",
            "name": "_data",
            "type": "bytes",
          },
        ],
        "name": "safeTransferFrom",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "operator",
            "type": "address",
          },
          {
            "internalType": "bool",
            "name": "approved",
            "type": "bool",
          },
        ],
        "name": "setApprovalForAll",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "bytes4",
            "name": "interfaceId",
            "type": "bytes4",
          },
        ],
        "name": "supportsInterface",
        "outputs": [
          {
            "internalType": "bool",
            "name": "",
            "type": "bool",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "symbol",
        "outputs": [
          {
            "internalType": "string",
            "name": "",
            "type": "string",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "tokenURI",
        "outputs": [
          {
            "internalType": "string",
            "name": "",
            "type": "string",
          },
        ],
        "stateMutability": "pure",
        "type": "function",
      },
      {
        "inputs": [],
        "name": "totalSupply",
        "outputs": [
          {
            "internalType": "uint256",
            "name": "",
            "type": "uint256",
          },
        ],
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "internalType": "address",
            "name": "from",
            "type": "address",
          },
          {
            "internalType": "address",
            "name": "to",
            "type": "address",
          },
          {
            "internalType": "uint256",
            "name": "tokenId",
            "type": "uint256",
          },
        ],
        "name": "transferFrom",
        "outputs": [],
        "stateMutability": "nonpayable",
        "type": "function",
      },
    ],
    "address": {
      "1": "0xaf0326d92b97df1221759476b072abfd8084f9be",
    },
    "name": "WagmiMintExample",
  },
]
`;

exports[`tryFetchProxyImplementation: fetches implementation ABI 1`] = `
[
  {
    "abi": [
      {
        "constant": false,
        "inputs": [
          {
            "name": "newImplementation",
            "type": "address",
          },
        ],
        "name": "upgradeTo",
        "outputs": [],
        "payable": false,
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "constant": false,
        "inputs": [
          {
            "name": "newImplementation",
            "type": "address",
          },
          {
            "name": "data",
            "type": "bytes",
          },
        ],
        "name": "upgradeToAndCall",
        "outputs": [],
        "payable": true,
        "stateMutability": "payable",
        "type": "function",
      },
      {
        "constant": true,
        "inputs": [],
        "name": "implementation",
        "outputs": [
          {
            "name": "",
            "type": "address",
          },
        ],
        "payable": false,
        "stateMutability": "view",
        "type": "function",
      },
      {
        "constant": false,
        "inputs": [
          {
            "name": "newAdmin",
            "type": "address",
          },
        ],
        "name": "changeAdmin",
        "outputs": [],
        "payable": false,
        "stateMutability": "nonpayable",
        "type": "function",
      },
      {
        "constant": true,
        "inputs": [],
        "name": "admin",
        "outputs": [
          {
            "name": "",
            "type": "address",
          },
        ],
        "payable": false,
        "stateMutability": "view",
        "type": "function",
      },
      {
        "inputs": [
          {
            "name": "_implementation",
            "type": "address",
          },
        ],
        "payable": false,
        "stateMutability": "nonpayable",
        "type": "constructor",
      },
      {
        "payable": true,
        "stateMutability": "payable",
        "type": "fallback",
      },
      {
        "anonymous": false,
        "inputs": [
          {
            "indexed": false,
            "name": "previousAdmin",
            "type": "address",
          },
          {
            "indexed": false,
            "name": "newAdmin",
            "type": "address",
          },
        ],
        "name": "AdminChanged",
        "type": "event",
      },
      {
        "anonymous": false,
        "inputs": [
          {
            "indexed": false,
            "name": "implementation",
            "type": "address",
          },
        ],
        "name": "Upgraded",
        "type": "event",
      },
    ],
    "address": {
      "1": "0xa0b86991c6218b36c1d19d4a2e9eb0ce3606eb48",
    },
    "name": "FiatToken",
  },
]
`;
